mixer/test/spybackend/session.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package spybackend
+
+import (
+	"context"
+
+	"fmt"
+	"net"
+
+	"google.golang.org/grpc"
+
+	adptModel "istio.io/api/mixer/adapter/model/v1beta1"
+	"istio.io/istio/mixer/template/listentry"
+	"istio.io/istio/mixer/template/metric"
+	"istio.io/istio/mixer/template/quota"
+)
+
+type (
+	server interface {
+		Addr() net.Addr
+		Close() error
+		Run()
+	}
+
+	sessionServer struct {
+		listener net.Listener
+		shutdown chan error
+		server   *grpc.Server
+		behavior *behavior
+		requests *requests
+	}
+)
+
+var _ adptModel.InfrastructureBackendServer = &sessionServer{}
+var _ metric.HandleMetricServiceServer = &sessionServer{}
+var _ listentry.HandleListEntryServiceServer = &sessionServer{}
+var _ quota.HandleQuotaServiceServer = &sessionServer{}
+
+// Non request time rpcs
+func (s *sessionServer) Validate(c context.Context, r *adptModel.ValidateRequest) (*adptModel.ValidateResponse, error) {
+	s.requests.validateRequest = append(s.requests.validateRequest, r)
+	return s.behavior.validateResponse, s.behavior.validateError
+}
+func (s *sessionServer) CreateSession(c context.Context, r *adptModel.CreateSessionRequest) (*adptModel.CreateSessionResponse, error) {
+	s.requests.createSessionRequest = append(s.requests.createSessionRequest, r)
+	return s.behavior.createSessionResponse, s.behavior.createSessionError
+}
+func (s *sessionServer) CloseSession(c context.Context, r *adptModel.CloseSessionRequest) (*adptModel.CloseSessionResponse, error) {
+	s.requests.closeSessionRequest = append(s.requests.closeSessionRequest, r)
+	return s.behavior.closeSessionResponse, s.behavior.closeSessionError
+}
+
+func (s *sessionServer) HandleMetric(c context.Context, r *metric.HandleMetricRequest) (*adptModel.ReportResult, error) {
```

Comment:
consider embedding `noSessionServer` inside `SessionServer`. That way it is clear what is same and what is different between the two.

Reply:
D